build-aux/snap/local/patch-dl.py

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+from elftools.elf.elffile import ELFFile
+
+
+parser = argparse.ArgumentParser()
+
+craft_prime = os.environ.get('CRAFT_PRIME')
+parser.add_argument('--prime', nargs=1, default=craft_prime, required=(craft_prime is None))
+
+craft_part_install = os.environ.get('CRAFT_PART_INSTALL')
+parser.add_argument('--install', nargs=1, default=craft_part_install, required=(craft_part_install is None))
+
+parser.add_argument('linker')
+
+args = parser.parse_args()
+
+def is_shared_exec(path):
+    with open(path, 'rb') as f:
+        if f.read(4) != b'\x7fELF':
+            return False
+        f.seek(0, 0)
+        elf = ELFFile(f)
+        for segment in elf.iter_segments(type='PT_INTERP'):
+            return True
+        return False
+
+owned_executables = []
+
+for dirpath, dirnames, filenames in os.walk(args.install):
+    for filename in filenames:
+        path = os.path.join(dirpath, filename)
+        if os.path.islink(path):
+            continue
+        if not is_shared_exec(path):
+            continue
+        rel = os.path.relpath(path, args.install)
+        prime_path = os.path.join(args.prime, rel)
+        install_st = os.lstat(path)
+        prime_st = os.lstat(path)
+        if install_st.st_dev != prime_st.st_dev:
+            continue
+        if install_st.st_ino != prime_st.st_ino:
+            continue
```

Comment:
Probably should comment here. We want to see what files in prime directory are owned by the specific part. So we look at hardlinks.

Though  thinking about it, since we break the hardlinks after, we should also probably deal with non hardlinked files.

Reply:
... in case we re-run the build.

Reply:
I have verified. `craftclt default` always re-introduce the hardlink when priming.